ql/src/java/org/apache/hadoop/hive/ql/exec/repl/bootstrap/load/IncrementalLoad.java

```diff
@@ -0,0 +1,307 @@
+      String location = dir.getPath().toUri().toString();
+      DumpMetaData eventDmd = new DumpMetaData(new Path(location), conf);
+
+      if (!shouldReplayEvent(dir, eventDmd.getDumpType(), dbName, tableName)) {
+        this.log.debug("Skipping event {} from {} for table {}.{} maxTasks: {}",
+                eventDmd.getDumpType(), dir.getPath().toUri(), dbName, tableName, maxTasks);
+        continue;
+      }
+
+      this.log.debug("Loading event {} from {} for table {}.{} maxTasks: {}",
+              eventDmd.getDumpType(), dir.getPath().toUri(), dbName, tableName, maxTasks);
+
+      // event loads will behave similar to table loads, with one crucial difference
+      // precursor order is strict, and each event must be processed after the previous one.
+      // The way we handle this strict order is as follows:
+      // First, we start with a taskChainTail which is a dummy noop task (a DependecyCollectionTask)
+      // at the head of our event chain. For each event we process, we tell analyzeTableLoad to
+      // create tasks that use the taskChainTail as a dependency. Then, we collect all those tasks
+      // and introduce a new barrier task(also a DependencyCollectionTask) which depends on all
+      // these tasks. Then, this barrier task becomes our new taskChainTail. Thus, we get a set of
+      // tasks as follows:
+      //
+      //                 --->ev1.task1--                          --->ev2.task1--
+      //                /               \                        /               \
+      //  evTaskRoot-->*---->ev1.task2---*--> ev1.barrierTask-->*---->ev2.task2---*->evTaskChainTail
+      //                \               /
+      //                 --->ev1.task3--
+      //
+      // Once this entire chain is generated, we add evTaskRoot to rootTasks, so as to execute the
+      // entire chain
+
+      MessageHandler.Context context = new MessageHandler.Context(dbName, tableName, location,
+              taskChainTail, eventDmd, conf, hive, driverContext.getCtx(), this.log);
+      List<Task<? extends Serializable>> evTasks = analyzeEventLoad(context);
+
+      if ((evTasks != null) && (!evTasks.isEmpty())) {
+        ReplStateLogWork replStateLogWork = new ReplStateLogWork(replLogger,
+                dir.getPath().getName(),
+                eventDmd.getDumpType().toString());
+        Task<? extends Serializable> barrierTask = TaskFactory.get(replStateLogWork);
+        for (Task<? extends Serializable> t : evTasks) {
+          t.addDependentTask(barrierTask);
+          this.log.debug("Added {}:{} as a precursor of barrier task {}:{}",
+                  t.getClass(), t.getId(), barrierTask.getClass(), barrierTask.getId());
+        }
+        this.log.debug("Updated taskChainTail from {}:{} to {}:{}",
+                taskChainTail.getClass(), taskChainTail.getId(), barrierTask.getClass(), barrierTask.getId());
+        taskChainTail = barrierTask;
+        maxTasks -= evTasks.size();
```

Comment:
This doesn't give the proper count as each tasks in evTasks may have childTasks too which should be included in the count. Shall use TaskTracker Utility which does this.

Reply:
done